set-web/src/main/java/eu/europeana/set/web/service/impl/UserSetServiceImpl.java

```diff
@@ -243,41 +248,67 @@ public int validatePosition(String position, List<String> items) throws Applicat
-	// validate position
-	// -1 if invalid
-	int positionInt = validatePosition(position, existingUserSet.getItems());
-
-	// build new item URL
 	String newItem = UserSetUtils.buildItemUrl(WebUserSetFields.BASE_ITEM_URL, datasetId, localId);
+    // check if the position is "pin" and is a EntityBestItem set then
+	// insert the item at the 0 position
+    if(StringUtils.equals(position, WebUserSetModelFields.PINNED_POSITION) && existingUserSet.isEntityBestItemsSet()) {
+		return insertItem(existingUserSet, newItem, 0, true);
+    } else {
+		// validate position
+		// -1 if invalid
+		int positionInt = validatePosition(position, existingUserSet.getItems(), existingUserSet.getPinned());
+		return insertItem(existingUserSet, newItem, positionInt, false);
+	}
+    }
 
-	// check if item already exists in the Set, if so remove it
-	// insert item to Set in the indicated position (or last position if no position
-	// was indicated).
-	UserSet extUserSet = null;
-	if (existingUserSet.getItems() == null) {
-	    addNewItemToList(existingUserSet, -1, newItem);
-	    extUserSet = updateItemList(existingUserSet);
-	} else {
-	    if (!existingUserSet.getItems().contains(newItem)) {
-		// add item
-		addNewItemToList(existingUserSet, positionInt, newItem);
-		extUserSet = updateItemList(existingUserSet);
-	    } else {
-		// replace item
-		int currentPos = existingUserSet.getItems().indexOf(newItem);
-		if (currentPos == positionInt) {
-		    // do not change user set, just add pagination
-		    // the items is already present at the correct position
-		    extUserSet = getUserSetUtils().updatePagination(existingUserSet);
+	/**
+	 * check if item already exists in the Set, if so remove it
+	 * insert item to Set in the indicated position (or last position if no position
+	 * was indicated).
+	 *
+	 * if pinnedItem : Then increase the counter by one while adding the item.
+	 * No need to update the counter while replacing the existing item
+	 *
+	 * @param existingUserSet
+	 * @param newItem
+	 * @param positionInt
+	 * @return
+	 */
+    private UserSet insertItem(UserSet existingUserSet, String newItem, int positionInt, boolean pinnedItem) {
+		UserSet extUserSet = null;
+		if (existingUserSet.getItems() == null) {
+			addNewItemToList(existingUserSet, -1, newItem);
+			// if pinned item, increase the pinned counter by 1
+			if(pinnedItem) {
+				existingUserSet.setPinned(existingUserSet.getPinned()+1);
+			}
+			extUserSet = updateItemList(existingUserSet);
 		} else {
-		    replaceItem(existingUserSet, positionInt, newItem);
-		    extUserSet = updateItemList(existingUserSet);
+			if (!existingUserSet.getItems().contains(newItem)) {
+				// add item
+				addNewItemToList(existingUserSet, positionInt, newItem);
+				// if pinned item, increase the pinned counter by 1
+				if(pinnedItem) {
+					existingUserSet.setPinned(existingUserSet.getPinned()+1);
+				}
+				extUserSet = updateItemList(existingUserSet);
+
+			} else {
+				// replace item
```

Comment:
need to verify if the repositioning is allowed for pinned items

Reply:
same as before just added the logic for the counter, if item is changed from pinned to normal or otherway around